Breeding-Insight/bi-api
# [BI-1255] Germplasm Cache

# Description
**Story:** (https://breedinginsight.atlassian.net/jira/software/c/projects/BI/boards/1?modal=detail&selectedIssue=BI-1255)

Is based onto BI-1056, so it might be best to wait for this review until that one is merged. 

Key points about germplasm cache: 
- Cache was written as a wrapper class so any DAO can implement it with relative ease. 
  - DAOs will just have to use the cache to fetch, post, and put and the updating of the cache will be handled by the cache class. 
- Added tests for the cache. The tests are dependent on thread sleeping to mimic waiting time for api calls. I tried to make it fairly robust and tried to avoid tweaking wait times to get tests to pass. 
- Key events of the cache: 
  - On micronaut startup, cache for all existing programs are initialized. 
    - A failed cache refresh will not interrupt startup, it will just log the error. The cache will be updated next time germplasm are pulled. 
    - If a GET request is made before micronaut is finished with the initial cache load for that program, the cache will wait for the current refresh to finish instead of making another GET request. 
  - On a failed refresh the cache for that program is invalidated so we aren't left with stale data. Cache is refreshed again next time germplasm are requested. 
  - On post, the cache is refreshed. The cache accounts for multiple posts in a single request process by queuing up refresh requests. If there is already a refresh queued up, another one won't be added to the queue because the waiting one will do the work of both. 
  - If there is existing data in the cache when the cache is refreshing, the existing data will be returned instead of having to wait for the refreshed data. 



# Dependencies
develop on bi-web and sgn.

# Testing
- Using the germplasm importer, import this file into a breedbase program [BI-1255_halfton_of_germplasm.xls](https://github.com/Breeding-Insight/bi-api/files/7848094/BI-1255_halfton_of_germplasm.xls)
- Go to the "All Germplasm" table and make sure all of the germplasm appear
- Now we have to get rid of the germplasm cache so we can see the speed difference between cache and non-cache. 
- Turn off bi-api and breedbase
- Turn bi-api back on and wait for it to be running
  - We do this first so the cache initialization on startup fails and we don't have germplasm cached for breedbase program.
- Turn breedbase back on and wait for it to be running
- Go to the "All Germplasm" page and notice how long it takes for the table to load (should be several seconds)
  - This is the time it takes to fetch germplasm from breedbase. 
- Now the cache is populated.
- Refresh the page, the table should display in less than a second, this is a success!



# Checklist:
